server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java

```diff
@@ -417,6 +419,22 @@ protected void onCompletion() {
                 }
             }
         }
+
+        @Override
+        public void onCancelled() {
+            try {
+                if (task instanceof CancellableTask t) {
+                    t.ensureNotCancelled();
+                }
```

Comment:
Since we're handling a cancellation in this method can we assert that the task is always a `CancellableTask` here?

```suggestion
                } else {
                    assert false : "task must be cancellable";
                }
```

Reply:
Oh wait I see, we're keeping hold of the exception for later. Forget I said anything...

Reply:
I like the assertion, I would prefer the assertion to be the first statement in the method and then continue assuming this is a `CancellableTask`. 
**Pros:** I see the assertion as a "guard" of the method that ensures everything is ok before moving forward with the execution. 
**Cons:** We cannot use `instanceof` pattern matching.

If you do not have an objection I will move forward with something like this:
```
            assert task instanceof CancellableTask : "task must be cancellable";
            try {
                ((CancellableTask) task).ensureNotCancelled();
            } catch (TaskCancelledException e) {
                nodeResponseTracker.discardIntermediateResponses(e);
            }
```

Reply:
Changed.